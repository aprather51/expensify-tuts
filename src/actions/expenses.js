import uuid from 'uuid';
import database from '../firebase/firebase'

//ADD_EXPENSE *NEW* (For Firebase)
export const addExpense = (expense) =>({
    type: 'ADD_EXPENSE',
    expense
});

//ADD_EXPENSE *OLD*
// export const addExpense = ({ 
//     description = '', 
//     note = '', 
//     amount = 0, 
//     createAt = 0
// } = {}) =>({
//     type: 'ADD_EXPENSE',
//     expense:{
//         id: uuid(),
//         description,
//         note,
//         amount,
//         createAt
//     }
// });



//** Setting up Expense data to dispatch based on *expenseData = {}*  **//
export const startAddExpense = (expenseData = {}) => {
    //Setting "return" only for redux-thunk... This function now get called interally by Redux and it get returned with argument called "dispatch"... 
    return (dispatch) => {
        //...And dispatch which gives an access inside items to write some data for firebase. 
        const { 
            description = '', 
            note = '', 
            amount = 0, 
            createAt = 0
        } = expenseData
        //set constant to select data from above that need to be pushed into firebase
        const expense = { description, note, amount, createAt }
    
        //firebase syntax to create/open array called 'expenses' and *push()* the variable called expense (see above "const expense = {...}) that contain items and values such as -- "description, note, amount and createAt" to firebase....The *then()* call back the success case for push to get called with "ref" together...
        return database.ref('expenses').push(expense).then((ref) => {
            //dispatching action called addExpense (Above) which will make changes when new data get updated.
            dispatch(addExpense({
                //dispatching a unique key with call name "ref"
                id: ref.key,
                //dispatching expense properties i.e. dispatch, note, amount, and createAt 
                ...expense
                //at end it pushes id based on .then(ref) and expense properities together.
            }));
        });
    };
};//End of Export
   

//REMOVE_EXPENSE
export const removeExpense = ({id} = {}) =>({
    type: 'REMOVE_EXPENSE',
    id
});

//
export const startRemoveExpense = ({id} = {}) => { 
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }));
        })
    };
};

//EDIT_EXPENSE
export const editExpense = (id, update) => ({
    type: 'EDIT_EXPENSE',
    id,
    update
});

//SET_EXPENSE
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//FETCH THE DATA AT START
export const startSetExpenses =  () => {
    /*Return with dispatch variable to get darta */
    return (dispatch) => {
        /* Return with promise string. Collect data reference called expenses from firebase with value 
        at once. When data is successfully obtained, it proceed to then() call with snapshot variable.
        The snapshot variable snaps at object structure.*/
        return database.ref('expenses').once('value').then((snapshot) => {
            
            /* setting up expenses with nothing */
            const expenses = [];

            /* Collect object structure and show list of items in array, Since snapshot isn't working
            best with redux. childSnapshot is get called as variable.*/
            snapshot.forEach((childSnapshot) => {
                /* To push() childSnapshot propties such as key and values to firebase*/
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            /* dispatching setExpense action with expense variable*/
            dispatch(setExpenses(expenses))
        });
       
    };
};