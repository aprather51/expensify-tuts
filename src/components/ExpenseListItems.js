import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItems = ({id, description, amount, createAt}) => (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description} </h3>
      </Link> 
      <p>
        {numeral(amount / 100).format('$0,0.00')} 
        - 
        {moment(createAt).format('dddd, MMMM D YYYY')}</p>
      <hr/>
    </div>
    
);

export default ExpenseListItems;