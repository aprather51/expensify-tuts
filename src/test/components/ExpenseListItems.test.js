import React from 'react'
import { shallow } from 'enzyme';
import ExpenseListItems from '../../components/ExpenseListItems';
import expenses from '../fixures/expenses';

test('Render ExpenseListItems with expenses', () => {
    const wrapper = shallow(<ExpenseListItems {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
});
