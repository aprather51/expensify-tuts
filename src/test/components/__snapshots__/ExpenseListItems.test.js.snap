// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ExpenseListItems with expenses 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Gum
       
    </h3>
  </Link>
  <p>
    $1.95
    -
    Wednesday, December 31 1969
  </p>
  <hr />
</div>
`;
